services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt

```diff
@@ -13,36 +14,67 @@ class HankeKayttajaService(
 
     @Transactional
     fun saveNewTokensFromApplication(applicationData: ApplicationData, hankeId: Int) {
-        // Contacts new for this application
-        val newContacts =
+        val contacts =
             applicationData
                 .customersWithContacts()
                 .flatMap { it.contacts }
-                .filter { !it.email.isNullOrBlank() && !it.name.isNullOrBlank() }
-
-        val existingEmails =
-            hankeKayttajaRepository
-                .findByHankeIdAndSahkopostiIn(hankeId, newContacts.map { it.email!! })
-                .map { it.sahkoposti }
-
-        val contactsToCreate =
-            newContacts
-                .filter { contact -> !existingEmails.contains(contact.email) }
-                .distinctBy { it.email }
-        contactsToCreate.forEach { createToken(hankeId, it) }
+                .mapNotNull { (name, email) ->
+                    when {
+                        name.isNullOrBlank() -> null
+                        email.isNullOrBlank() -> null
+                        else -> UserContactDto(name, email)
+                    }
+                }
```

Comment:
I think flatter is better, even though you need `!!` in a couple of places.
```suggestion
                .filterNot { it.email.isNullOrBlank() || it.name.isNullOrBlank() }
                .map { (name, email) -> UserContactDto(name!!, email!!) }
```

Reply:
I also think this would be clearer without the destructuring.

Reply:
Done as suggested.